geozero/src/feature_processor.rs

```diff
@@ -6,22 +6,46 @@ use crate::property_processor::PropertyProcessor;
 #[allow(unused_variables)]
 pub trait FeatureProcessor: GeomProcessor + PropertyProcessor {
     /// Begin of dataset processing
+    ///
+    /// ## Invariants
+    ///
+    /// - `dataset_begin` is called _only once_ for an entire dataset.
+    /// - `dataset_begin` is called before all other methods, including `feature_begin`,
+    ///   `properties_begin`, `geometry_begin`, and all methods from [`GeomProcessor`] and
+    ///   [`PropertyProcessor`]
     fn dataset_begin(&mut self, name: Option<&str>) -> Result<()> {
         Ok(())
     }
     /// End of dataset processing
+    ///
+    /// ## Invariants
+    ///
+    /// - `dataset_end` is called _only once_ for an entire dataset.
+    /// - No other methods may be called after `dataset_end`.
     fn dataset_end(&mut self) -> Result<()> {
         Ok(())
     }
     /// Begin of feature processing
+    ///
+    /// - `idx` refers to the positional row index in the dataset. For the `n`th row, `idx` will be
```

Comment:
```suggestion
    /// - `idx` the positional row index in the dataset. For the `n`th row, `idx` will be
```

nit: seems like a couple wasted words (here and duplicated elsewhere)